### ISSUE TITLE
Fatal Python error: none_dealloc (Struct, datetime)

### ISSUE THREAD

**Issue by Ev3** [jcrist/msgspec#408]
### Description

Hi all. 
I have such an error wile using your amazing lib)) It's related to **Struct**, **datetime** and GC

Code to reproduce:

```
import datetime
import msgspec

class AStruct(msgspec.Struct):
    dt_val: datetime.datetime


data_str = msgspec.json.encode(
    {
        "dt_val": datetime.datetime.now()
    }
)

for _ in range(10000):
    msgspec.json.decode(data_str, type=AStruct)
```

```
Fatal Python error: none_dealloc: deallocating None
Python runtime state: initialized

Current thread 0x00007f71e9b0d740 (most recent call first):
  File "/app/test.py", line 20 in <module>

Extension modules: msgspec._core (total: 1)
```

**Spec:**
Linux x64
Python: 3.10.11 and 3.11.3 (all from dockerhub, bullseye's version)
msgspec:  from pypi version "==0.15.0"

Thanks!

**Comment by jcrist** [comment]
Thanks for the excellent reproducible issue, this has been resolved in #409.

This regression was introduced in 0.15.0, for now you can use 0.14.2. I'll try to cut a patch release later today with the fix.

**Comment by Ev3** [comment]
Thanks a lot. 
0.14.2 works like a charm for now))

**Comment by jcrist** [comment]
This has now been released as part of 0.15.1.
